pkg/raft/raft.go

```diff
@@ -1049,7 +1051,11 @@ func (r *raft) campaign(t CampaignType) {
 		term = r.Term
 		voteMsg = pb.MsgVote
 	}
-	r.logger.Infof("%x is starting a new %selection at term %d", r.id, pre, term)
+	if hung {
```

Comment:
Why do we need to call out these cases?

Is it necessarily "hung"? What if the candidate voluntarily/forcefully starts a new campaign, without waiting for a timeout? For example, using `RawNode.Campaign()` method.

Also, there is [this](https://github.com/cockroachdb/cockroach/blob/c64fb80786b82b0a848bf59e271d8b2ce44db674/pkg/raft/raft.go#L1653-L1654) happy case transition from pre-candidate to candidate when we've won the pre-election. Would it be logged as "hung" here?

Maybe there is a way to frame this message more generally so that it doesn't necessarily imply an election timeout? But then the question is how much value would this message add. Do you want to specifically catch cases of "hung" election, or just want to extend logging with more context? In the latter case, would printing some extra state like `r.state` achieve the effect?

Reply:
If you specifically target the "hung" elections, maybe there is some place up the stack where it can be detected unambiguously (`tickElection`?).